ccf_specs.md

```diff
@@ -161,6 +161,91 @@ This specification uses CDDL notation to express CBOR data items:
 - `[* Foo]`: zero or more Foo in CBOR array
 - `#6.nnn(type)`: CBOR tag with tag number nnn and content type of "type".
 
+## Serialization Considerations
+
+CCF is a data format that uses a subset of CBOR with additional requirements for validity and deterministic encoding.
+
+### Cadence Types and Values Encoding
+
+[Cadence values have types](https://developers.flow.com/cadence/language/values-and-types).  For example:
+- `true` has the type `Bool` 
+- `"hello"` has the type `String` 
+- `let aos: [String] = ["hello", "world"]` declares `aos` of type `[String]`. 
+- `let aoa: [AnyStruct] = ["hello", true]` declares `aoa` of type `[AnyStruct]`.
+
+CCF encoding decouples value encoding from type encoding.  For example:  
+- Cadence booleans is encoded as a tuple of type and value: the `Bool` type and its value.  
+- Cadence strings is encoded as a tuple of type and value: the `String` type and its value.
+
+For compactness, CCF encoding skips type encoding when type can be inferred.
+
+As one example, the type of each element in `[String]` can be inferred to have type `String`.  Given this, CCF encodings of `[String]` can skip encoding each element's type of `String`.  Instead, CCF encodes `[String]` as a tuple like this: `[String]` type and a value representing a list of element values.  This avoids redundantly encoding each element's type.
+
+If type cannot be inferred, it must be encoded as part of a tuple representing a type and its value.  For example, the type of each element in `[AnyStruct]` cannot be inferred.  The type `AnyStruct` is the top type of all non-resource types and array elements can be of heterogenous types. In this case, `[AnyStruct]` is encoded as a tuple like this: `[AnyStruct]` type and a value being a list of tuples representing each element's type and value.
```

Comment:
Maybe avoid the use of the terms "infer" and "inference", as it might be confused with type inference in the sense of the Cadence type checker, which, based on the types of the values infers the element type.

Here, an encoder rather omits type information for element values, if the value's type matches the outer container's element type. For example, for
- `aos`: When encoding the first value, `"example"`, which has type `String`: The value type matches the outer element type `String` => no type info for value needed
- `aoa`: When encoding the first value, `"example"`, which has type `String`: The value type does not match the outer element type `AnyStruct` => type info for value need